### PR TITLE
fix(billing): enforce the retention window on evaluation runs

### LINKED ISSUE #1987: Evaluations date range is not plan-gated, bypassing retention entitlement
## Summary

The Evaluations page date-range dropdown lets any user pick 30d / 60d / 90d regardless of subscription plan, bypassing the retention gate every other trace-telemetry surface enforces. The page renders the shared `DateFilterSelect` **without** the `retentionDays` prop that drives the lock, so the lock can never fire — and the evaluations backend applies no retention clamp either, so wider ranges genuinely return data rather than being merely cosmetically selectable.

This is a wiring defect, not an unbuilt feature: the gate exists, works, and is applied on Traces, Sessions, Users, Detectors and Dashboards. Evaluations is the one surface never passed the plan state.

## How to replicate

1. Sign in as a workspace on **Free** (retention cap 15 days) or **Starter** (30 days).
2. Open Traces and its date-range dropdown: presets beyond the plan window render with a lock icon and an "N-day retention limit · click to upgrade" tooltip, and clicking routes to the upgrade flow.
3. Open the **Evaluations** page and its date-range dropdown.
4. **Observed:** every preset including 30d/60d/90d is freely selectable — no lock, no tooltip, no upgrade routing, on any plan.
5. Select 90d and confirm the runs list actually returns rows older than the plan's retention window. The wider range is honored end-to-end, not just displayed.

## Root cause (confirmed)

The lock is entirely prop-driven and Evaluations never supplies the prop; the backend has no clamp either, so nothing enforces the cap at either layer.

1. **The lock returns "unlocked" when `retentionDays` is absent.** `isOptionLocked()` short-circuits to `false` for *every* option when it is null/undefined (`frontend/ui/src/lib/date-filter.ts:38-45`, specifically `:42`). It is consumed once per option at `frontend/ui/src/components/date-filter-select.tsx:88`; a locked option renders the lock icon (`:121`) and routes the click to `onUpgradeClick` instead of applying the range (`:100-121`). With no prop, `locked` is always `false` and the gate is inert.
2. **The preset list reaches 90d.** `DATE_FILTER_OPTIONS` includes 30d, 60d, 90d (`date-filter.ts:17-32`), and the comment at `:26-28` states 60d/90d are *meant* to render locked on lower tiers.
3. **Evaluations passes no plan state.** `frontend/ui/src/features/evaluations/views/evaluations-view.tsx:312-321` passes only `dateFilter` / `customStartDate` / `customEndDate` / `onDateFilterChange` / `onCustomRangeChange` — no `retentionDays`, no `onUpgradeClick` — and never imports `useRetention`.
4. **Every other surface wires it.** Traces `app/projects/[projectId]/traces/page.tsx:208`, Sessions `:132`, Users `:133`, Detectors list `:154`, Detector detail `:232`, Dashboard `:157` all pass `retentionDays={retention.retentionDays}`; Traces additionally wires `onUpgradeClick` (`traces/page.tsx:208-209`).
5. **The caps are real and fail closed.** `RETENTION_DAYS` (`frontend/packages/core/src/ee/billing/plans.ts:151-171`) is Free **15**, Starter **30**, Pro **90**, Enterprise **null**; `getRetentionDays()` falls back to `FAIL_CLOSED_RETENTION_DAYS = 15` (`:159`) for unknown plans. `frontend/ui/src/lib/hooks/use-retention.ts` reads `workspace.billingPlan`, defaults to FREE, and returns `undefined` while loading — so nothing locks transiently on surfaces that do wire it.

**Not merely cosmetic:** retention is also enforced server-side, so on gated surfaces a hand-crafted wide request is still clamped. `backend/rest/retention.py` (`clamp_retention_window` pulls `start_after` forward; `enforce_retention_by_time` 403s by-id lookups) is wired into `backend/rest/routers/traces.py:93,207,267,340,362` plus sessions, users, dashboards, detectors and live SSE, and mirrored in `frontend/ui/src/lib/server/retention.ts`. The evaluations path has **none** of it: `frontend/ui/src/app/api/projects/[projectId]/evaluations/runs/route.ts` contains no retention or clamp handling. Note this Next route is the **only** server-side reader of evaluation runs — the SDK-facing gateway `backend/rest/routers/public/eval.py` is write-only for runs (its `_UPSTREAM_ROUTES` allowlist registers `evaluation-runs` as POST-only; its sole GETs are `datasets` / `dataset-versions`, which take no time window), so a `GET` for runs is refused at routing and never forwarded. One clamp, in the Next route, is therefore sufficient.

**Secondary follow-up (not part of this fix):** the dashboard widget-builder *preview* window is also ungated — `frontend/ui/src/features/dashboards/components/WidgetBuilderPage.tsx:296` uses raw `RANGE_PRESETS` (`features/dashboards/range-presets.ts:11`) with no retention filter. It is a preview control, not the live dashboard query (whose own filter *is* gated at `dashboard/[dashboardId]/page.tsx:157`), so impact is limited.

## Impact

- **Retention cap not enforced:** workspaces on lower-retention plans can query evaluation runs across the full 90-day window, in the UI *and* against the backend.
- **Inconsistent behaviour:** the same control shows locks and upgrade prompts on every other surface but is wide open here, so the retention limit reads as arbitrary.
- **No backstop:** because the eval path applies no clamp, nothing server-side catches the ungated UI — the wider data is actually served.

## Acceptance criteria

- The Evaluations `DateFilterSelect` receives `retentionDays` and `onUpgradeClick` from `useRetention`, exactly like the other six surfaces, so presets beyond the plan window render locked and route to upgrade.
- On Free, 30d/60d/90d lock; on Starter, 60d/90d lock; on Pro/Enterprise all presets stay selectable — matching `RETENTION_DAYS`.
- The evaluations runs endpoint enforces the same server-side clamp as the telemetry routes, reusing the existing helpers rather than a parallel mechanism — so a wider range is clamped rather than served even against a hand-crafted request. This means `.../evaluations/runs/route.ts`, the only reader; `public/eval.py` is write-only for runs and has nothing to clamp, which should be pinned by a test so a future ungated read cannot be added silently.
- An **omitted or unparseable** `started_after` must be clamped to the cutoff too — an unbounded request must not be a wider query than explicitly asking for 90 days.
- Nothing locks transiently while plan state is loading (`retentionDays` `undefined` → no locks), consistent with the other surfaces.
- Plan retention numbers are unchanged and the fail-closed default for unknown plans is preserved.
- (Secondary) the widget-builder preset preview is either gated the same way or has a documented reason for exemption.

### PR DESCRIPTION
**Pull request by dark-sorceror**
## Summary

Fixes: #1987

The Evaluations date-range dropdown let any workspace pick 30d / 60d / 90d regardless of plan, and the runs endpoint actually served the wider window. Retention gating is built and enforced on every other telemetry surface — Traces, Sessions, Users, Detectors (list + detail), Dashboards — in two halves: the UI hands `DateFilterSelect` a `retentionDays` so out-of-window presets render locked, and the read endpoint clamps `start_after` server-side so a hand-crafted request cannot out-reach the plan either. Evaluations had **neither** half.

## UI — pass the plan window to the control

`frontend/ui/src/features/evaluations/views/evaluations-view.tsx:164,329-330,440` now mirrors the other six surfaces: `useRetention(projectId)` supplies `retentionDays` and `onUpgradeClick`, and the `PricingDialog` that flow opens is mounted. Without the dialog `onUpgradeClick` would be a dead control — `useRetention` only flips its own `showPricing` state; the surface owns the rendering, as on all six gated pages (e.g. `traces/page.tsx:395`).

No plan numbers changed. The window still comes from `RETENTION_DAYS` (`frontend/packages/core/src/ee/billing/plans.ts:151-156` — Free 15 / Starter 30 / Pro 90 / Enterprise null) through `getRetentionDays()`, which fails closed to 15 for an unrecognized plan.

## Server — clamp the runs window

`frontend/ui/src/app/api/projects/[projectId]/evaluations/runs/route.ts:85-100` reuses `clampStartAfter` (`lib/server/retention.ts`), the same helper the detector proxies use, resolving the plan the same way:

```ts
const billingPlan = workspace?.billingPlan || PlanType.FREE;
const startedAfter = parseDateParam(
  clampStartAfter(billingPlan, searchParams.get("started_after")),
);
```

**Lists clamp, they do not 403** — the established split: `clamp_retention_window` pulls `start_after` forward for lists while `enforce_retention_by_time` 403s by-id lookups (`backend/rest/retention.py:54-84`). Two properties make this airtight rather than decorative:

- **Omitting `started_after` is not a bypass.** The helper substitutes the cutoff for a missing *or unparseable* bound. Before this change, dropping the parameter was the widest query available.
- **The clamp lands on both query branches.** The route splits on sort field — `startedAt`/`status` sort in the DB, `cost`/`elapsedMs` in Node (`route.ts:53-59`) — but both build the same `where`, so clamping the parsed bound covers both. Pinned by its own test.

The 1-hour boundary buffer is inherited from the helper and matches `retention.py`, so "Last 15 days" on Free does not race the server clock.

## Python: nothing to clamp, asserted rather than assumed

The other candidate site for the clamp is `backend/rest/routers/public/eval.py`. **That router has no read of evaluation runs to clamp** — it is the SDK-facing write gateway, and its `_UPSTREAM_ROUTES` allowlist registers every `evaluation-runs` shape as POST-only. Its only GETs are `datasets` / `dataset-versions`, which take no time window, so `GET /api/v1/public/evaluation-runs?started_after=…` is refused at routing with 405 and never forwarded. The Next route is the sole reader, so one clamp suffices.

`tests/rest/test_public_eval_gateway_security.py::TestNoUngatedRetentionRead` makes that executable rather than a claim in a description — asserting the allowlist is write-only and reads are refused before any forward, with a failure message naming `clamp_retention_window` as the prerequisite for adding one. **A guard, not a regression test:** it passes on `main` too, and exists to fail the day someone adds a listing GET the Next route never sees.

## Tests

**`evaluations-retention.test.tsx`** (new). `date-filter.test.ts` and `date-filter-select.test.tsx` already cover the lock logic and the control — both in isolation, so neither can see whether the view *passes* the prop, which is the whole defect. This mounts the real view over the real `useRetention` and reads the trigger label. Full matrix: Free locks 30/60/90; Starter locks 60/90; Pro and Enterprise lock nothing; an unknown plan locks like Free; `retentionDays === undefined` (plan loading) locks nothing.

**`runs/route.test.ts`** — 11 new cases: per-plan clamping, Pro/Enterprise pass-through, in-window pass-through, unbounded and unparseable bounds, fail-closed for unknown plan and missing workspace row, the cost/elapsed sort branch, and `started_before` untouched.

Each layer verified by reverting only its own source file (`git stash push -- <path>`):

```
evaluations-view.tsx reverted → Tests  4 failed | 4 passed (8)
runs/route.ts reverted       → Tests  7 failed | 17 passed (24)
     → expected 3884400001 to be less than 5000      # 44.96d = 60d − (15d+1h), Free
     → expected 2588400000 to be less than 5000      # 29.96d = 60d − (30d+1h), Starter
     → expected undefined to be an instance of Date  # unbounded + unparseable
```

The passing cases in each are the must-not-over-lock guards (Pro, Enterprise, loading, in-window), which hold in both directions by design.

After: `Test Files 211 passed (211)` / `Tests 1985 passed (1985)` (full `frontend/ui` vitest); `1559 passed, 5 skipped` (`uv run pytest tests/ -q`); ruff, ESLint and Prettier clean. `tsc --noEmit` reports six errors, all pre-existing on `main` in untouched files; CI does not run it.

## Deliberately not included

**The widget-builder preview presets** (`WidgetBuilderPage.tsx:548`) — flagged as secondary in #1987; a different defect with a different fix. It does not serve out-of-window data: the builder renders its own `DropdownMenu` over `RANGE_PRESETS` rather than `DateFilterSelect`, so there is no prop to pass, and its preview query is already clamped server-side (`backend/rest/routers/dashboards.py:144`). Picking 90d on Free returns 15 days under a "Last 90 days" label — misleading, not privileged. A wrinkle for that change: `handleRangePick` writes the preset into the shared project date-filter store (`:122-125`), which the dashboard and trace list re-clamp, so the user's stored selection silently changes underneath them. A proper fix filters the presets *and* refuses to persist an out-of-window id.

**By-id reads of evaluation runs.** `.../evaluations/runs/[runId]` and the compare route have no `enforce_retention_by_time` analogue. With the list clamped there is no UI path to an out-of-window run id, so this is not part of the reported bug, but it is the same shape and the natural follow-up.

---
## Summary by cubic
Plan-gates the Evaluations date range and clamps the runs list to each workspace’s retention window to close an entitlement leak. Previously any workspace could pick 30/60/90 days and the API served it; now out‑of‑window presets are locked, an over‑wide pick collapses once the plan resolves, and the server bounds started_after.

- UI: `EvaluationsView` passes `retentionDays`/`onUpgradeClick` from `useRetention` to `DateFilterSelect`, mounts `PricingDialog`, and passes `hasSubscription` to route upgrades correctly. Derives the active range with `clampDateFilter` so a pre‑load 90d pick collapses to the plan window when retention lands. Uses `toPlanType` in `@traceroot/core` to normalize unknown `billingPlan` values to Free.
- API: `GET /api/projects/[projectId]/evaluations/runs` clamps `started_after` using `workspace.billingPlan` (fallback Free). Missing/unparseable bounds resolve to the cutoff; `started_before` is unchanged. Clamp applies on both sort branches; Free 15d, Starter 30d, Pro 90d, Enterprise unlimited; includes the standard 1‑hour buffer.
- Public gateway: `evaluation-runs` remains write‑only; tests assert no ungated read path and note that any future read must apply `rest.retention.clamp_retention_window`.
- Tests: cover UI locking, applied preset → `started_after`, collapse after retention resolves, per‑plan clamping and edge cases, and billing helpers (plan normalization and subscription state).

<sup>Written for commit 079339f8df9d5d3e65fe0370294274bae6d67f96. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/traceroot-ai/traceroot/pull/1991?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: frontend/ui/src/features/evaluations/views/evaluations-view.tsx [hunk 1]
```diff
@@ -426,6 +436,13 @@ function RunsTab({ projectId }: { projectId: string }) {
           isDeleting={del.isPending}
         />
       )}
+
+      <PricingDialog
+        open={retention.showPricing}
+        onOpenChange={retention.closePricing}
+        workspaceId={retention.workspaceId}
+        currentPlan={(retention.billingPlan as PlanType) || PlanType.FREE}
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply frontend/ui/src/features/evaluations/views/evaluations-view.tsx]: Resolved in ce2545b8

Valid — `billing_plan` is a free-form TEXT column (no CHECK constraint), so an unrecognized value is reachable, and raw it makes `getPlanOrder` return undefined, which makes `isUpgrade` false for every plan and labels all four CTAs "Downgrade". Rather than inline the check I added `toPlanType(plan)` to core's plans.ts next to the existing fail-closed `getRetentionDays` (it mirrors `normalize_plan` in backend/shared/config.py) and used it here; unit-tested in core plus a view test asserting an unknown plan renders Free as "Current Plan" with no Downgrade CTA. Note the same raw cast exists at 8 other call sites (traces/sessions/users/detectors/dashboard pages, billing settings, RetentionGateBanner, SidebarUpgradeButton) — migrating those is a follow-up, out of scope for this PR.

File: frontend/ui/src/features/evaluations/evaluations-retention.test.tsx [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+      </ToastProvider>
+    </QueryClientProvider>,
+  );
+}
+
+/** Mount, open the preset popover, and click `label`. */
+async function pickPreset(label: string) {
+  mount();
+  fireEvent.click(await screen.findByRole("button", { name: DEFAULT_LABEL }));
+  fireEvent.click(screen.getByRole("button", { name: label }));
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply frontend/ui/src/features/evaluations/evaluations-retention.test.tsx:83]: Resolved in 3eae2ba0

Valid — the label-only assertions passed even when the picked window never reached the server. The helpers now decode `started_after` off the mocked fetch URL: `expectApplied(label, days)` waits for the latest `/evaluations/runs` request to carry that window, and `expectLocked` asserts the request is still on the 14-day default. Confirmed discriminating with two mutations of the view — dropping `started_after` from the query fails 7 tests, and pinning the bounds to `"14d"` regardless of the picked preset fails 5; the old label-only version passed both.
